honeybadgermpc/hbavss.py

```diff
@@ -229,57 +232,144 @@ def __exit__(self, typ, value, traceback):
-            if avss_msg == HbAVSSMessageType.OK and sender not in ok_set:
-                ok_set.add(sender)
-            elif avss_msg == HbAVSSMessageType.READY and sender not in ready_set:
-                ready_set.add(sender)
-
-        multicast(HbAVSSMessageType.READY)
+        r1_set = set()
+        r2_set = set()
+        ready_sent = False
+        r1_sent = False
+        r2_sent = False
+        r1_phi = [None] * self.n
+        r2_phi = [None] * self.n
 
-        # if 2t+1 ready -> output shares
-        while len(ready_set) < 2 * self.t + 1:
-            sender, avss_msg = await recv()  # First value is the `sid`
-            if avss_msg == HbAVSSMessageType.READY and sender not in ready_set:
+        while True:
+            # main recv loop for Bracha-style agreement and implicate handling
+            sender, avss_msg = await recv()
+            # OK
+            if avss_msg[0] == HbAVSSMessageType.OK and sender not in ok_set:
+                ok_set.add(sender)
+            # READY
+            if avss_msg[0] == HbAVSSMessageType.READY and sender not in ready_set:
                 ready_set.add(sender)
-
-        return shares
+            # IMPLICATE
+            if avss_msg[0] == HbAVSSMessageType.IMPLICATE:
+                # Handling Implication
+                j = sender
+                j_pk = avss_msg[1]
+                j_k = avss_msg[2]
+                # discard if PKj ! = g^SKj
+                if self.public_keys[j] != pow(self.g, j_pk):
+                    logger.debug("[%d] invalid implicate from sender %s",
+                                 self.my_id, sender)
+                    continue
+                # decrypt and verify
+                implicate_msg = await avid.retrieve(tag, j)
+                j_z = loads(implicate_msg)
+                j_shared_key = pow(ephemeral_public_key, j_pk)
+                j_share, j_aux, j_witnesses = SymmetricCrypto.decrypt(
+                    str(j_shared_key).encode("utf-8"), j_z[j_k])
+                if self.poly_commit.verify_eval(
+                        commitments[j_k], sender+1, j_share, j_aux, j_witnesses):
+                    continue
+
+                # proceed to share recovery
+                logger.debug("[%d] Share recovery activated by %d", self.my_id, j)
+                # setup the coordinates for interpolation
+                c_coords = [None] * secret_size
+                phi_coords = [None] * secret_size
+                aux_coords = [None] * secret_size
+                w_coords = [None] * secret_size
+                for i in range(secret_size):
+                    c_coords[i] = (i, commitments[i])
+                    phi_coords[i] = (i, shares[i])
+                    aux_coords[i] = (i, auxes[i])
+                    w_coords[i] = (i, witnesses[i])
+                # interpolate commitments
+                interpolated_c = [None] * self.n
+                for i in range(self.n):
```

Comment:
This should not be using lagrange - we should switch to Vandermonde/FFT. I think we got rid of lagrange from most places in our code.

Reply:
@tyurek

Reply:
I'm not aware of how we would do FFT interpolation for commitments in G1. If someone wants to tackle it, feel free, but I don't think it's terribly important as this is just fault handling anyway. It would be new work with little benefit

Reply:
interpolation in a finite field should use the polynomial class and FFT and all that though